securedrop/journalist.py

```diff
@@ -575,18 +575,38 @@ def download_single_submission(sid, fn):
 @app.route('/reply', methods=('POST',))
 @login_required
 def reply():
+    msg = request.form['msg']
+
+    # Reject empty replies
+    if not msg:
+        flash("You cannot send an empty reply!", "error")
+        return redirect(url_for('col', sid=g.sid))
+
     g.source.interaction_count += 1
     filename = "{0}-{1}-reply.gpg".format(g.source.interaction_count,
                                           g.source.journalist_filename)
-    crypto_util.encrypt(request.form['msg'],
+    crypto_util.encrypt(msg,
                         [crypto_util.getkey(g.sid), config.JOURNALIST_KEY],
                         output=store.path(g.sid, filename))
     reply = Reply(g.user, g.source, filename)
-    db_session.add(reply)
-    db_session.commit()
 
-    flash("Thanks! Your reply has been stored.", "notification")
-    return redirect(url_for('col', sid=g.sid))
+    try:
+        db_session.add(g.source)
```

Comment:
>I think as written this will throw an InvalidRequestError because the underlying source has been deleted. 

Only in scenario I described where one journalist deletes a source while another journalist is writing a reply to that source.

>But stepping back a bit: also recall that adding back a source doesn't regenerate the source directory / source GPG key that will be deleted when the source is deleted, so this line can be removed.

So while I wrote this just to be explicit, it's actually harmful in either (i) the case described above or (ii) the same case as above except where a source deletes their own account instead of another journalist. In either of these cases as you point out the source will get added back to the database, but the necessary and complimentary filesystem actions executed during `Source` initialization will not be performed. Not only does this screw up system state, it partially un-deletes an account and associated data that is supposed to be deleted.

My explicitness is not necessary as items returned by queries already present in the database are automatically tracked in the session without needed to be explicitly added:

> add() is used to place instances in the session. For transient (i.e. brand new) instances, this will have the effect of an INSERT taking place for those instances upon the next flush. For instances which are persistent (i.e. were loaded by this session), they are already present and do not need to be added.

-- http://docs.sqlalchemy.org/en/latest/orm/session_basics.html#adding-new-or-existing-items